src/vpn-api/utils/openvpn.ts

```diff
@@ -0,0 +1,318 @@
+import VpnConnector = require('telnet-openvpn');
+
+import { Netmask } from './netmask';
+
+const parsePossibleInt = (s: string): number | string => {
+	if (/^[1-9]\d*$/.test(s)) {
+		return parseInt(s, 10);
+	}
+	return s;
+};
+
+export interface VpnClientUntrustedData {
+	username: string;
+}
+export interface VpnClientTrustedData {
+	common_name: string;
+	ifconfig_pool_remote_ip: string;
+	trusted_ip: string;
+	trusted_port: number;
+}
+export interface VpnClientBytecountData {
+	bytes_received: number;
+	bytes_sent: number;
+}
+// CLIENT:CONNECT
+export interface VpnClientConnectData extends VpnClientUntrustedData {
+	password: string;
+}
+// CLIENT:ESTABLISHED
+export interface VpnClientEstablishedData
+	extends VpnClientUntrustedData,
+		VpnClientTrustedData {}
+// CLIENT:DISCONNECT
+export interface VpnClientDisconnectData
+	extends VpnClientUntrustedData,
+		Partial<VpnClientTrustedData>,
+		Partial<VpnClientBytecountData> {
+	time_duration: number;
+}
+
+export const isTrusted = (data: any): data is VpnClientTrustedData =>
+	data.common_name != null;
+export const hasBytecountData = (data: any): data is VpnClientBytecountData =>
+	data.bytes_received != null && data.bytes_sent != null;
+
+const VpnLogLevels = {
+	D: 'debug',
+	I: 'info',
+	n: 'notice',
+	W: 'warning',
+	N: 'error',
+	F: 'emerg',
+} as const;
+type ValueOf<T> = T[keyof T];
+
+export declare interface VpnManager {
+	on(event: 'process:error', callback?: (err: Error) => void): this;
+	on(
+		event: 'process:exit',
+		callback?: (code?: number, signal?: string) => void,
+	): this;
+	on(event: 'manager:connect', callback?: () => void): this;
+	on(event: 'manager:data', callback?: (data: string) => void): this;
+	on(
+		event: 'log',
+		callback?: (level: ValueOf<typeof VpnLogLevels>, message: string) => void,
+	): this;
+	on(
+		event: 'client:connect',
+		callback?: (
+			clientId: number,
+			keyId: number,
+			data: VpnClientConnectData,
+		) => void,
+	): this;
+	on(
+		event: 'client:address',
+		callback?: (clientId: number, address: string, primary: boolean) => void,
+	): this;
+	on(
+		event: 'client:established',
+		callback?: (clientId: number, data: VpnClientEstablishedData) => void,
+	): this;
+	on(
+		event: 'client:bytecount',
+		callback?: (clientId: number, data: VpnClientBytecountData) => void,
+	): this;
+	on(
+		event: 'client:disconnect',
+		callback?: (clientId: number, data: VpnClientDisconnectData) => void,
+	): this;
+}
+
+export class VpnManager extends EventEmitter {
+	private process?: ChildProcess;
+	private readonly connector = new VpnConnector();
+	private buf?: string;
+
+	constructor(
+		private instanceId: number,
+		private vpnPort: number,
+		private mgtPort: number,
+		private subnet: Netmask,
+		private gateway?: string,
+	) {
+		super();
+		// proxy `data` events from connector, splitting at newlines
+		this.connector.connection.on('data', data => {
+			const lines = ((this.buf || '') + data.toString()).split(/\r?\n/);
```

Comment:
If you start with
```
private buf = ''
```
then you can make this just
```suggestion
			const lines = (this.buf + data.toString()).split(/\r?\n/);
```

Reply:
Unfortunately `Array<T>.pop` returns `T | undefined` so the `|| ''` remains necessary.